pkg/remotes/docker/discoverer.go

```diff
@@ -0,0 +1,375 @@
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package docker
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/reference"
+	"github.com/containerd/containerd/remotes"
+	"github.com/containerd/containerd/remotes/docker"
+	"github.com/docker/docker/errdefs"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	artifactspec "github.com/oras-project/artifacts-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/net/context/ctxhttp"
+)
+
+func WithDiscover(ref string, resolver remotes.Resolver, opts *docker.ResolverOptions) (remotes.Resolver, error) {
+	opts = NewOpts(opts)
+
+	r, err := reference.Parse(ref)
+	if err != nil {
+		return nil, err
+	}
+
+	return &dockerDiscoverer{
+		header:     opts.Headers,
+		hosts:      opts.Hosts,
+		refspec:    r,
+		reference:  ref,
+		repository: strings.TrimPrefix(r.Locator, r.Hostname()+"/"),
+		tracker:    docker.NewInMemoryTracker(),
+		Resolver:   resolver}, nil
+}
+
+type dockerDiscoverer struct {
+	hosts      docker.RegistryHosts
+	header     http.Header
+	refspec    reference.Spec
+	reference  string
+	repository string
+	tracker    docker.StatusTracker
+	remotes.Resolver
+}
+
+var localhostRegex = regexp.MustCompile(`(?:^localhost$)|(?:^localhost:\\d{0,5}$)`)
+
+func (d *dockerDiscoverer) filterHosts(caps docker.HostCapabilities) (hosts []docker.RegistryHost, err error) {
+	h, err := d.hosts(d.refspec.Hostname())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, host := range h {
+		if host.Capabilities.Has(caps) || localhostRegex.MatchString(host.Host) {
+			hosts = append(hosts, host)
+		}
+	}
+
+	return hosts, nil
+}
+
+func (d *dockerDiscoverer) Discover(ctx context.Context, desc ocispec.Descriptor, artifactType string) ([]artifactspec.Descriptor, error) {
+	ctx = log.WithLogger(ctx, log.G(ctx).WithField("digest", desc.Digest))
+
+	hosts, err := d.filterHosts(docker.HostCapabilityResolve)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hosts) == 0 {
+		return nil, errdefs.NotFound(errors.New("no discover hosts"))
+	}
+
+	ctx, err = docker.ContextWithRepositoryScope(ctx, d.refspec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	v := url.Values{}
+	v.Set("artifactType", artifactType)
+	query := "?" + v.Encode()
+
+	var firstErr error
+	for _, originalHost := range hosts {
```

Comment:
Can you explain a little bit why the code needs to query the referrer from multiple hosts?

Reply:
This is how the docker resolver works.